File: Java-CodeCoverage/src/tech/getrealgood/util/Strings.java
package tech.getrealgood.util;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 *
 */
public final class Strings {

    /**
     * @param s
     * @return
     */
    public static String toUpper(String s) throws StringsException {

        if (Objects.isNull(s))
            throw new StringsException("Null parameter passed in for s");
        return s.toUpperCase();

    }

    /**
     *
     * @param s
     * @return
     * @throws StringsException
     */
    public static List<String> toUpper(List<String> s) throws StringsException {

        if (Objects.isNull(s))
            throw new StringsException("Null parameter passed in for s");

        var result = new ArrayList<String>(s.size());

        for (var s1: s) {

            result.add(toUpper(s1));
        }

        return result;

    }

    /**
     * @param s
     * @return
     */
    public static String toLower(String s) throws StringsException {
        if (Objects.isNull(s))
            throw new StringsException("Null parameter passed in for s");
        return s.toLowerCase();

    }

    /**
     *
     * @param s
     * @return
     * @throws StringsException
     */
    public static List<String> toLower(List<String> s) throws StringsException {

        if (Objects.isNull(s))
            throw new StringsException("Null parameter passed in for s");

        var result = new ArrayList<String>(s.size());

        for (var s1: s) {

            result.add(toLower(s1));
        }

        return result;

    }

    /**
     * @param s
     * @param token
     * @param trim
     * @return
     */
    public static List<String> split(String s, char token, boolean trim) throws StringsException {
        if (Objects.isNull(s))
            throw new StringsException("Null parameter passed in for s");

        if(s.length() == 0)
            return new ArrayList<>(0);

        List<String> result = Arrays.<String>asList(s.split(String.valueOf(token)));

        if (trim) {
            result = result.stream().map(String::trim).collect(Collectors.toList());
        }

        return result;
    }

    /**
     * @param s
     * @param token
     * @param trim
     * @return
     */
    public static String concat(List<String> s, String token, boolean trim) throws StringsException {

        if (Objects.isNull(s))
            throw new StringsException("Null parameter passed in for s");

        String result = new String("");
        int counter = 0;
        int length = s.size() - 1;

        for (var e : s) {

            if (trim) {
                if(e.length() == 0)
                    continue;
                result += e.trim();
            }
            else {
                if(e.length() == 0)
                    continue;
                result += e;
            }

            if (Objects.isNull(token))
                continue;
            else {
                if (counter < length) {
                    result += token;
                    counter++;
                }

            }

        }

        return result;
    }

    /**
     * @param s
     * @param orig
     * @param w
     * @param trim
     * @return
     * @throws StringsException
     */
    public static String replace(String s, char orig, char w, boolean trim) throws StringsException {
        if (Objects.isNull(s))
            throw new StringsException("Null value passed in for s");

        if (trim)
            s = s.trim();

        char[] chars = s.toCharArray();

        for (var i = 0; i < chars.length; i++)
            if (chars[i] == orig)
                chars[i] = w;

        return new String(chars);
    }

    /**
     * @param s
     * @param orig
     * @param w
     * @return
     */
    public static String replace(String s, char orig, char w) throws StringsException {
        if (Objects.isNull(s))
            throw new StringsException("Null value passed in for s");
        return replace(s, orig, w, false);
    }

}

File: Typescript/__files__/strings.ts
export class Strings {

    public static toUpper(o: string) : string {
        if(o == null)
            throw new StringsError("o == null");
        if(o.length == 0)
            return o;
        return o.toUpperCase();
    }

    public static toLower(o: string) : string {
        if(o == null)
            throw new StringsError("o == null");
        if(o.length == 0)
            return o;
        return o.toLowerCase();
    }

    public static split(o: string, token : string, trim : boolean) : string[] {
        if(o == null)
            throw new StringsError("o == null");

        if(token == null)
            throw new StringsError("token == null");

        if(token.length == 0)
            throw new StringsError("token.length == 0");

        if(o.length == 0)
            return [o];

        if(trim)
            o = o.trim();

        let result = o.split(token);

        if (trim) {
            for (let i in result) {
                i = i.trim();
            }
        }

        return result;

    }

    public static concat(o: string[], token : string, trim : boolean) : string {

        if(o == null)
            throw new StringsError("o == null");

        if(o.length == 0)
            throw new StringsError("o.length == 0");

        if(token == null)
            throw new StringsError("token == null");

        let result = "";

            for (let i of o) {
                if(trim) i = i.trim();
                result += i + token;
            }

         return result.substring(0, result.length - token.length);

    }

    public static replace(o: string, token1 : string, token2 : string, trim : boolean) : string {

        if(o == null)
            throw new StringsError("o == null");

        if(o.length == 0)
            return o;

        if(token1 == null)
            throw new StringsError("token1 == null");

        if(token1.length == 0)
            throw new StringsError("token1.length == 0");

        if(token2 == null)
            throw new StringsError("token2 == null");

        if(trim)
            o = o.trim();

        while (o.indexOf(token1) != -1) {
            o = o.replace(token1, token2);
        }
        return o;

    }

}

export class StringsError {

    constructor(s : string) {
        this.message = s;
    }

    message : string
}